Babylon/commands/api/payloads/upsert.py

```diff
@@ -0,0 +1,127 @@
+            return CommandResponse.fail()
+
+        if _match:
+            result = data_o
+            for item in data_stdin:
+                if item['reportId'] != old['reportId']:
+                    continue
+                d = dict()
+                d[key] = [item]
+                test = flatten(d, separator=".")
+                for k, v in test.items():
+                    i = k.index("0")
+                    end = k[i + 2:]
+                    nk = f"{key}.{idx}.{end}"
+                    data_o_f.update({nk: v})
+                result = unflatten_list(data_o_f, separator=".")
+
+    if query and not flat and not (dashboard_view or scenario_view):
+        _match = jmespath.search(query, data_o)
+        if _match is None:
+            logger.info("Query error: object not found")
+            return CommandResponse.fail()
+        key = _match.get("key")
+        old = _match.get("value")
+        try:
+            idx = data_o[key].index(old)
+        except AttributeError as exp:
+            logger.warn(exp)
+            return CommandResponse.fail()
+
+        if _match:
+            result = data_o
+            for item in data_stdin:
+                if item['id'] != old['id']:
+                    continue
+                d = dict()
+                d[key] = [item]
+                test = flatten(d, separator=".")
+                for k, v in test.items():
+                    i = k.index("0")
+                    end = k[i + 2:]
+                    nk = f"{key}.{idx}.{end}"
+                    data_o_f.update({nk: v})
+                result = unflatten_list(data_o_f, separator=".")
+
+    if not query and not flat and not (dashboard_view or scenario_view):
+        for item in enumerate(data_stdin):
+            _q = f"{section}[?id=='{item['id']}'] |"
+            _q += "{" + "\"key\": '{s}'".format(s=section) + ", value: [0]}"
+            _match = jmespath.search(_q, data_o)
+            key = _match.get("key") if _match else section
+            found = _match.get("value") if _match else False
+            if found:
+                idx = data_o[key].index(_match.get("value"))
+                d = dict()
+                d[key] = [item]
+                test = flatten(d, separator=".")
+                for k, v in test.items():
+                    i = k.index("0")
+                    end = k[i + 2:]
+                    nk = f"{key}.{idx}.{end}"
+                    data_o_f.update({nk: v})
+                result = unflatten_list(data_o_f, separator=".")
+            else:
+                data_o[key] = data_o[key] or []
+                data_o[key].append(item)
+                result = data_o
+
+    if not query and flat:
+        for item in data_stdin:
+            result = get_section_and_replace(section, item, data_o)
+
+    with target_file.open(mode='w') as _f:
+        yaml_loader.dump(result, target_file)
+        logger.info(f"Content dumped in {_f.name}")
+    return CommandResponse.success()
```

Comment:
Ok this function is a tough one ^^ !
I have seen many redundancy in here and maybe we could refacto this
Here is a quick example of what could be done on it (caution: it hasnt been tested !)
```python
    yaml_loader = YAML()
    data_o = yaml.safe_load(origin_file.open())
    data_o_f = flatten(yaml.safe_load(origin_file.open()), separator=".")
    data_s = click.get_text_stream('stdin', encoding="utf-8")
    data_stdin = yaml.safe_load(data_s.read())
    result = None

    if query and not flat and (dashboard_view or scenario_view):
        selector = "dashboardsView" if dashboard_view else "scenarioview"
        result = _get_results(query, data_o, data_stdin, data_o_f, "reportId", selector)

    if query and not flat and not (dashboard_view or scenario_view):
        result = _get_results(query, data_o, data_stdin, data_o_f, "id")

    if not query and not flat and not (dashboard_view or scenario_view):
        for item in enumerate(data_stdin):
            _q = f"{section}[?id=='{item['id']}'] |"
            _q += "{" + "\"key\": '{s}'".format(s=section) + ", value: [0]}"
            _match = jmespath.search(_q, data_o)
            key = _match.get("key") if _match else section
            found = _match.get("value") if _match else False
            if found:
                idx = data_o[key].index(_match.get("value"))
                result = _normalize(key, item, idx, data_o_f)
            else:
                data_o[key] = data_o[key] or []
                data_o[key].append(item)
                result = data_o

    if not query and flat:
        for item in data_stdin:
            result = get_section_and_replace(section, item, data_o)

    with target_file.open(mode='w') as _f:
        yaml_loader.dump(result, target_file)
        logger.info(f"Content dumped in {_f.name}")
    return CommandResponse.success()


def _normalize(key, item, idx, data_o_f):
    d = dict()
    d[key] = [item]
    test = flatten(d, separator=".")
    for k, v in test.items():
        i = k.index("0")
        end = k[i + 2:]
        nk = f"{key}.{idx}.{end}"
        data_o_f.update({nk: v})
    return unflatten_list(data_o_f, separator=".")


def _get_result_if_match(data_o, data_stdin, key, idx, data_o_f, old, key_stdin):
    result = data_o
    for item in data_stdin:
        if item[key_stdin] != old[key_stdin]:
            continue
        result = _normalize(key, item, idx, data_o_f)
    return result


def _get_results(query, data_o, data_stdin, data_o_f, params, selector=""):
    _match = jmespath.search(query, data_o)
    if _match is None:
        logger.info("Query error: object not found")
        return CommandResponse.fail()
    key, old = _match.get("key"), _match.get("value")

    if selector:
        selected_dict = data_o["webApp"]["options"]["charts"][selector]
    else:
        selected_dict = data_o[key]

    try:
        idx = selected_dict.index(old)
    except AttributeError as exp:
        logger.warn(exp)
        return CommandResponse.fail()

    if _match:
        return _get_result_if_match(data_o, data_stdin, key, idx, data_o_f, old, params)
```
Remember, this is an example, in order to show you how we could 'simplify' the code

Reply:
sure, i agree :)